feat/14 JWT + OAuth2 인증 시스템 구현 및 토큰 재발급 API 추가

seseseseo opened the pull request:
## 📌 관련 이슈
- Closes #536

---

## ✨ 기능 요약

| 항목        | 내용 |
|-------------|------|
| 🆕 기능명    | JWT + OAuth2 기반 인증 시스템 구현 |
| 🔍 목적      | JWT 기반 Access/Refresh Token 관리 및 소셜 로그인 연동을 통해 인증 흐름 구현 |
| 🛠️ 변경사항 | 인증 필터, 보안 설정, 리프레시 토큰 발급 및 저장, OAuth2 로그인 처리 로직 추가 |

---

## 📝 상세 내역

| 번호 | 내용 |
|------|------|
| 1️⃣ | `JwtTokenAuthenticationFilter` 구현 및 SecurityFilterChain에 OAuth2, JwtFilter 등록 |
| 2️⃣ | `TokenService` 및 `RefreshTokenService` 분리 및 책임 명확화 (토큰 발급 vs 저장소 관리) |
| 3️⃣ | Redis 기반 Refresh Token 저장/조회/삭제 기능 구현 |
| 4️⃣ | `OAuth2UserCustomService`, `OAuth2SuccessHandler`, `OAuth2Attributes` 구현으로 소셜 로그인 처리 |
| 5️⃣ | 토큰 재발급 API (`POST /api/auth/reissue`) 추가 |
| 6️⃣ | `User.updateSocialInfo()` 리팩토링으로 소셜 사용자 업데이트 방식 개선 |
| 7️⃣ | 관련 DTO (`TokenResponseDto`, `RefreshTokenRequestDto`) 추가 및 유효성 검증 어노테이션 적용 |

---

## ✅ 테스트 체크리스트
- [x] Access/Refresh Token 발급 및 재발급 정상 작동
- [x] Redis에 Refresh Token 저장 및 조회 정상 동작
- [x] Google, Kakao 소셜 로그인 성공 및 사용자 등록 확인
- [x] Refresh Token 없이 재발급 요청 시 예외 발생 확인
- [x] 회원가입, 로그인, 재발급 전 API 연동 완료
- [x] User 정보 업데이트 로직 정상 작동 확인

---

## 📸 포스트맨 캡처 사진
- [추가 예정] `/auth/reissue` 요청/응답 확인 스크린샷

---

## 🙋‍♀️ 리뷰어 참고사항
- 현재까지 JWT 발급/ 검증
- RefreshToken 관리
- Token 재발급 API
- OAuth2 로그인 (Google, Kakao) 연동
- OAuth2Attributes 기반 User 저장/ 업데이트 하는 회원 정보 매핑
- JWT 필터 설정 
등을 했습니다.

아직 회원가입 로직은 구현 전입니다. 또 JWT + OAuth2 가 실제로 작동하는지 테스트 하기 전입니다. 
이전 PR과 오늘 올린 PR은 회원가입/로그인을 위한 필수 파일들을 구성하는 PR입니다

seseseseo committed:
chore: 프로젝트 설정 파일 수정 (build.gradle, application.yml 등)

- Test 의존성 추가
- Redis 의존성 추가
- jwt, oauth 관련 application.yml 설정 추가

seseseseo committed:
feat:JWT 기반 인증 필터 및 보안 설정 구현

- JwtProperties 생성
- JwtUtils로 Access/Refresh Token 발급 및 검증 기능 구현
- JwtAuthenticationFilter 작성
- SecurityFilterChain에서 JWT 필터 적용 및 불필요한 보안 설정 비활성화
  (csrf, session, formLogin, logout 등)
- 인증이 필요한 경로(/api/**), 예외 경로(/api/auth/**) 설정
- BCryptPasswordEncoder Bean 등록

seseseseo committed:
feat: Redis 기반 리프레시 토큰 저장 및 재발급 로직 구현

- RedisConfig 설정 클래스 추가
- RefreshTokenService에서 RedisTemplate을 활용한 토큰 저장/조회/삭제 로직 구현
- TokenResponseDto 작성 (AccessToken, RefreshToken 포함)
- TokenService에서 리프레시 토큰 유효성 검증 및 토큰 재발급 기능 구현
- ErrorStatus 인증 부분 추가

seseseseo committed:
refactor: redisTemplate 타입 수정

- <String, String> 을 <String , Object> 로 바꿨습니다

seseseseo committed:
refactor: user 엔티티에서 직접 userDetail 구현하던 것 분리

- 보안 정책이 변경되면 엔티티에도 영향이 갈 것 같아서 UserDetails 를 분리함

seseseseo committed:
chore: enum 타입 패키지 구조 정리

- user domain 아래로 옮겼습니다.

seseseseo committed:
feat: OAuth2 구현을 위한 기본 클래스 작성

- provider 사용자 매핑을 위한 OAuth2AttributesDto 추가
- provider 에서 사용자를 로드하고 저장/ 업데이트 하는 OAuth2UserCustomService 구현
- 쿠키의 인증 요청을 관리하는 OAuth2AuthorizationRequestBasedOnCookieRepository 작성
- 쿠키 작업을 위한 CookieUtil 작성
- 성공 핸들러 작성

seseseseo committed:
feat: OAuth2 구현을 위한 securityFilterChain 설정 추가

seseseseo committed:
feat: AccessToken 발급용 API 구현

seseseseo committed:
chore : 패키지 구조 수정, UserDetailService 리네임 -> CustomUserDetailsService

seseseseo committed:
refactor : 잘못된 메서드 네이밍 수정, user 엔티티에 필드 추가
- JwtUtils 에서 AccessToken, RefreshToken 생성 메서드 네이밍이 잘못된 것을 수정
- user 엔티티에 민간, 공공에 필요한 필드 추가

seseseseo committed:
refactor: 잘못된 패키지 import 수정

seseseseo committed:
refactor: 전반적인 코드 리팩토링

- 기능은 이전과 같습니다

seseseseo committed:
chore: 프로젝트 설정 파일 수정 (application.yml)

seseseseo committed:
feat: 전략 패턴 기반 회원가입 서비스 구현

- UserRole(IND, GOV, NGO)에 따라 입력 필드가 다른 회원가입 전략 패턴 적용
- SignupService를 통해 역할에 따라 다른 입력값 처리
- SignupStrategy 인터페이스 및 각 역할별 구현체(IndSignupStrategy, GovSignupStrategy 등) 작성
- 입력 DTO 및 응답 DTO 분리 설계 (SignupRequestDto, SignupResponseDto)
- 회원가입 단일 엔드포인트(/api/auth/signup) 구현 및 유효성 검증 적용

seseseseo committed:
refactor: SecurityConfig 파일 리팩토링 (기능 수정X)

seseseseo committed:
feat: 로그인 구현

rabitis99 commented on src/main/java/com/example/emergencyassistb4b4/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,74 @@
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/auth")
```
api 설정에서 바로 들어갈거라서 빼서도 됩니다.

seseseseo committed:
chore: 프로젝트 설정 파일 수정 (build.gradle, application.yml 등)

- Test 의존성 추가
- application.yml 에 redirect-uri 부분 수정

seseseseo committed:
refactor: 전반적인 코드 리팩토링

rabitis99 commented on src/main/java/com/example/emergencyassistb4b4/auth/controller/AuthController.java:
logout 기능이 없네요

rabitis99 commented on src/main/java/com/example/emergencyassistb4b4/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,76 @@
+    public ResponseEntity<ApiResponse<TokenResponseDto>> signup(@Valid
+                                                         @RequestBody SignUpRequestDto requestDto, ServletRequest servletRequest) {
+        TokenResponseDto token = signUpService.signup(requestDto);
+        return ResponseEntity.ok(ApiResponse.onSuccess(SuccessStatus.SIGNUP_SUCCESS, token).getBody());
```
ApiResponse.onSuccess(SuccessStatus.SIGNUP_SUCCESS, token)으로만 처리하셔도 괜찮을 겁니다

seseseseo replied:
네 감사합니다

rabitis99 commented on src/main/java/com/example/emergencyassistb4b4/auth/oauth/handler/OAuth2SuccessHandler.java:
```diff
@@ -0,0 +1,96 @@
+    @Override
+    public void onAuthenticationSuccess(HttpServletRequest request, HttpServletResponse response,
+                                        Authentication authentication) throws IOException {
+        System.out.println("🔥 Success handler 호출됨");
```
System.out.println("🔥 Success handler 호출됨"); 확인 때문에 넣어 주신거죠?

seseseseo replied:
맞습니다 프론트 연결 하고있어서 이곳저곳에 sout 으로 확인 중이었는데 
파일들이 워낙 많아서 코드 확인을 못했네요 감사합니다

seseseseo committed:
test: 회원가입 단위 테스트, 서비스 통합 테스트 코드 작성

- 단위 : "카카오 로그인으로 개인 사용자가 가입 요청을 보내면, 중복 이메일이 없을 경우 사용자 정보가 저장되고 토큰이 발급된다."
- 단위 : "일반 로그인으로 개인 사용자가 가입 요청을 보내면, ~ 이하 동일"

- 통합 : " 일반 회원가입 성공 및 토큰 발급 "

Prototype-Km reviewed (approved):
고생하셨습니다

classseoha commented on src/main/java/com/example/emergencyassistb4b4/user/domain/User.java:
```diff
@@ -12,90 +11,73 @@
     private Long id;
 
+    @Column(name = "nickname", unique = true)
+    private String nickname;
```
oauth 가입 때문에 그냥 name이 아니라 nickname으로 가는걸까요??

seseseseo replied:
네 맞습니다 oauth 때문에 닉네임으로 쓰고있습니다
개인정보 동의 항목에서 이름은 불가능하고 
닉네임  profile_nickname.  
카카오 계정 account_email 
이렇게 2개 가져와서 쓰려고 합니다

seseseseo committed:
Merge branch 'dev/backend' into feat/14-oauth2

classseoha reviewed (approved):
주석 자세하게 달아주셔서 감사합니다 ㅎㅎ

seseseseo committed:
refactor: 역직렬화 문제 해결

- 프론트와 연동 중 어느부분에서 에러가 났는지 확인하기 위해 로그코드가 많이 남아있는 상태입니다.
- 다시 코드 깔끔하게 리팩토링 후 다시 커밋할 예정입니다.

seseseseo committed:
Merge remote-tracking branch 'origin/feat/14-oauth2' into feat/14-oauth2

seseseseo committed:
Merge branch 'dev/backend' into feat/14-oauth2

rabitis99 commented on src/main/java/com/example/emergencyassistb4b4/auth/signup/strategy/SignUpStrategy.java:
![image](https://github.com/user-attachments/assets/b001f5cb-47e2-4924-a9c1-b90408df4882)

rabitis99 commented on src/main/java/com/example/emergencyassistb4b4/global/config/WebOAuth2SecurityConfig.java:
```diff
@@ -0,0 +1,109 @@
+                        }))
+                .build();
+    }
+
```
@Bean
    public CorsConfigurationSource corsConfigurationSource() {
        CorsConfiguration config = new CorsConfiguration();
        config.setAllowedOriginPatterns(List.of("http://localhost:5173", "http://127.0.0.1:5173"));
        config.setAllowedMethods(List.of("GET", "POST", "PUT", "DELETE", "OPTIONS"));
        config.setAllowedHeaders(List.of("*"));
        config.setExposedHeaders(List.of("Authorization")); // 필요한 헤더 노출
        config.setAllowCredentials(true);

        UrlBasedCorsConfigurationSource source = new UrlBasedCorsConfigurationSource();
        source.registerCorsConfiguration("/**", config);  // 모든 경로에 대해 적용
        return source;
    }
이렇게 webconfig를 해야하는 걸로 알고 있습니다.

rabitis99 commented on src/main/java/com/example/emergencyassistb4b4/global/config/WebConfig.java:
```diff
@@ -0,0 +1,18 @@
+import org.springframework.web.servlet.config.annotation.CorsRegistry;
+import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;
+
+@Configuration
```
spring security 에서  안되는 걸로 알고 있습니다.

rabitis99 commented on src/test/resources/application-test.yml:
```diff
@@ -0,0 +1,18 @@
+spring:
```
.env 에서 관리해주세요

rabitis99 reviewed (approved):
주석, sout,controller의 api/는 금요일 변경하면 될 것같습니다.

seseseseo committed:
커밋 되돌리기

seseseseo committed:
refactor: 수정사항 반영

seseseseo committed:
refactor: 수정사항 반영

khg9900 commented on src/main/java/com/example/emergencyassistb4b4/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,65 @@
+
+    @PostMapping("/logout")
+    public ResponseEntity<ApiResponse<Void>> logout(ServletRequest servletRequest) {
+        String token = jwtUtils.resolveToken(servletRequest);
```
Service 내부에서 처리하는 것이 좋을 것 같아요!

khg9900 commented on src/main/java/com/example/emergencyassistb4b4/auth/dto/LoginRequestDto.java:
```diff
@@ -0,0 +1,23 @@
+    private UserRole userRole;
+
+    private LoginType loginType;
+
```
UserRole , LoginType 도 `@NotNull` 처리 해주시면 좋을 것 같아요!

rabitis99 replied:
Message도 추가 해야됩니다

khg9900 commented on src/main/java/com/example/emergencyassistb4b4/auth/dto/LoginResponseDto.java:
```diff
@@ -0,0 +1,12 @@
+
+public record LoginResponseDto(
+        String email,
+        String password,
```
응답에서는 비밀번호 제외 해주세요!

khg9900 commented on src/main/java/com/example/emergencyassistb4b4/auth/dto/SignUpRequestDto.java:
```diff
@@ -0,0 +1,38 @@
+    private String email;
+
+    @NotBlank(message = "비밀번호는 필수 입력 값입니다.")
+    private String password; // local만 필수
```
`@Pattern` 도 적용해주시면 좋을 것 같아요!

rabitis99 replied:
이건 패턴은 추후에 해도 괜찮을 것 같습니다. 까먹지 않게 주석만 달아도 괜찮을 것습니다

khg9900 commented on src/main/java/com/example/emergencyassistb4b4/auth/dto/LoginRequestDto.java:
dto 패키지 내부에 request, response 패키지 추가해서 구분해주시면 더 보기 좋을 것 같아요!

khg9900 commented on src/main/java/com/example/emergencyassistb4b4/user/domain/LoginType.java:
```diff
@@ -1,4 +1,4 @@
-package com.example.emergencyassistb4b4.user.enums;
+package com.example.emergencyassistb4b4.user.domain;
```
enum을 domain 패키지로 옮기신 이유를 여쭤봐도 될까요?

khg9900 commented on src/main/java/com/example/emergencyassistb4b4/auth/oauth/dto/OAuth2Attributes.java:
```diff
@@ -0,0 +1,65 @@
+
+    public static OAuth2Attributes of(String provider, Map<String, Object> attributes) {
+        return switch (provider.toLowerCase()) {
+            case "google" -> ofGoogle(attributes);
```
```
case GOOGLE -> ofGoogle(attributes);
```
처럼 enum으로 변경하면 더 좋을 것 같습니다!

khg9900 commented on src/main/java/com/example/emergencyassistb4b4/auth/oauth/dto/OAuth2Attributes.java:
```diff
@@ -0,0 +1,65 @@
+    private final String name;
+    private final String email;
+    private final Map<String, Object> attributes;
+    private final String provider;
```
여기도 enum으로 관리하면 좋을 것 같아요!

khg9900 commented on src/main/java/com/example/emergencyassistb4b4/auth/oauth/dto/OAuth2Attributes.java:
```diff
@@ -0,0 +1,65 @@
+ */
+@Getter
+public class OAuth2Attributes {
+    private final String name;
```
필드 선언 순서 아래랑 맞춰주시면 좋을 것 같습니다!

khg9900 commented on src/main/java/com/example/emergencyassistb4b4/auth/oauth/handler/OAuth2SuccessHandler.java:
```diff
@@ -0,0 +1,97 @@
+        clearAuthenticationAttributes(request, response);
+
+        //4. 리다이렉트
+        String targetUrl = getTargetUrl(tokens.accessToken());
```
쿼리파라미터에 Access Token을 넣는 것은 보안상 권장되지 않는 방식이라고 합니다!
쿠키로 전달하는 방식으로 변경하는 것을 추천드려요!

khg9900 commented on src/main/java/com/example/emergencyassistb4b4/auth/oauth/repository/OAuth2AuthorizationRequestBasedOnCookieRepository.java:
```diff
@@ -0,0 +1,84 @@
+    /**
+     * 요청에서 쿠키에 저장된 OAuth2 인증 요청 정보를 로드하고
+     * 삭제된 객체를 반환합니다.
+     * 현재 구현은 단순히 로드만 하고 쿠키를 실제로 삭제하지 않습니다.
```
내부에 removeAuthorizationRequestCookies 메서드가 있는데 혹시 주석 수정이 안 된 건가요?

khg9900 commented on src/main/java/com/example/emergencyassistb4b4/global/util/CookieUtil.java:
```diff
@@ -0,0 +1,85 @@
+
+        } catch (IOException | ClassNotFoundException e) {
+
+            throw new IllegalArgumentException("역직렬화 중 오류 발생", e);
```
ApiException으로 변환 부탁드립니다!

khg9900 commented on src/main/java/com/example/emergencyassistb4b4/global/util/CookieUtil.java:
```diff
@@ -0,0 +1,85 @@
+    // 객체를 직렬화해서 쿠키의 값으로 변환
+    public static String serialize(Object obj) {
+        if (!(obj instanceof Serializable)) {
+            throw new IllegalArgumentException("객체는 Serializable을 구현해야 합니다.");
```
ApiException 변환 부탁드립니다!

khg9900 commented on src/main/java/com/example/emergencyassistb4b4/global/util/CookieUtil.java:
```diff
@@ -0,0 +1,85 @@
+            try(ObjectInputStream ois = new ObjectInputStream(new ByteArrayInputStream(decoded))) {
+                Object obj = ois.readObject();
+                if (!cls.isInstance(obj)) {
+                    throw new IllegalArgumentException("역직렬화된 객체 타입이 일치하지 않음.");
```
ApiException 변환 부탁드립니다!

khg9900 commented on src/main/java/com/example/emergencyassistb4b4/global/status/ErrorStatus.java:
컨벤션에 맞게 수정 부탁드립니다!

khg9900 commented on src/main/java/com/example/emergencyassistb4b4/auth/service/LogoutService.java:
```diff
@@ -0,0 +1,31 @@
+    public void logout(String token) {
+        // 1. token 유효성 검사
+        if (!jwtUtils.validateToken(token)) {
+            throw new ApiException(ErrorStatus.INVAlID_ACCESS_TOKEN);
```
오타 있습니다! `INVAlID` → `INVALID`

khg9900 commented on src/main/java/com/example/emergencyassistb4b4/auth/token/RedisService.java:
```diff
@@ -0,0 +1,21 @@
+        stringRedisTemplate.delete("refresh:" + email);
+    }
+    public void addToBlackList(String token, long ttlMillis) {
+        stringRedisTemplate.opsForValue().set("blacklist:" + token, ttlMillis + "ms");
```
확인 한 번 해주시면 감사하겠습니다!
<img width="854" alt="스크린샷 2025-06-13 오후 2 03 00" src="https://github.com/user-attachments/assets/cc458fdd-77cb-4d53-b98f-fc46d7590cc5" />

khg9900 commented on src/main/java/com/example/emergencyassistb4b4/auth/token/TokenService.java:
```diff
@@ -0,0 +1,64 @@
+        refreshTokenService.saveToken(userId, newRefreshToken); // refresh 토큰 저장
+
+        // 6. 리프레시 토큰 저장 ( 기존 것을 덮어씀)
+        refreshTokenService.saveToken(userId, newRefreshToken);
```
```
refreshTokenService.saveToken(userId, newRefreshToken); // refresh 토큰 저장

        // 6. 리프레시 토큰 저장 ( 기존 것을 덮어씀)
        refreshTokenService.saveToken(userId, newRefreshToken);
```

동일한 코드가 연달아 있어서 하나는 삭제해도 될 것 같아요!
